connectors/byoc.py

```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
-            "indexed_document_count": indexed_count,
-            "completed_at": iso_utc(self.completed_at),
-        }
+    @property
+    def pipeline(self):
+        return Pipeline(self.get("connector", "pipeline"))
 
-        if exception is None:
-            self.status = JobStatus.COMPLETED
-            job_def["error"] = None
-        else:
-            self.status = JobStatus.ERROR
-            job_def["error"] = str(exception)
+    @property
+    def terminated(self):
+        return self.status in (JobStatus.ERROR, JobStatus.COMPLETED, JobStatus.CANCELED)
 
-        job_def["status"] = e2str(self.status)
+    @property
+    def indexed_document_count(self):
+        return self.get("indexed_document_count", default=0)
 
-        return await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+    @property
+    def indexed_document_volume(self):
+        return self.get("indexed_document_volume", default=0)
 
-    async def suspend(self):
-        self.status = JobStatus.SUSPENDED
-        job_def = {"status": e2str(self.status)}
+    @property
+    def deleted_document_count(self):
+        return self.get("deleted_document_count", default=0)
 
-        await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+    @property
+    def total_document_count(self):
+        return self.get("total_document_count", default=0)
 
-    @classmethod
-    def transform_filtering(cls, filtering):
-        # deepcopy to not change the reference resulting in changing .elastic-connectors filtering
-        filtering = (
-            {"advanced_snippet": {}, "rules": []}
-            if (filtering is None or len(filtering) == 0)
-            else deepcopy(filtering)
+    async def validate_filtering(self, validator):
+        validation_result = await validator.validate_filtering(self.filtering)
+
+        if validation_result.state != FilteringValidationState.VALID:
+            raise InvalidFilteringError(
+                f"Filtering in state {validation_result.state}. Expected: {FilteringValidationState.VALID}."
+            )
+        if len(validation_result.errors):
+            raise InvalidFilteringError(
+                f"Filtering validation errors present: {validation_result.errors}."
+            )
+
+    async def claim(self):
+        doc = {
+            "status": e2str(JobStatus.IN_PROGRESS),
+            "started_at": iso_utc(),
+            "last_seen": iso_utc(),
+            "worker_hostname": socket.gethostname(),
+        }
+        await self.index.update(doc_id=self.id, doc=doc)
+
+    async def done(self, ingestion_stats={}, connector_metadata={}):
+        await self._terminate(
+            JobStatus.COMPLETED, None, ingestion_stats, connector_metadata
         )
 
-        # extract value for sync job
-        filtering["advanced_snippet"] = filtering.get("advanced_snippet", {}).get(
-            "value", {}
+    async def fail(self, message, ingestion_stats={}, connector_metadata={}):
```

Comment:
Should we call this `error` as the names of `cancel` and `suspend` also adhere to the corresponding `JobStatus`?

Reply:
`error` is used by another method:

https://github.com/elastic/connectors-python/blob/d7053db73522da993bd4d32e074af4012030c13f/connectors/byoc.py#L208-L210

It's not required to use the same method name as the job status.

Reply:
Just thought it would be a bit cleaner/easier to read. But totally fine to keep it like that 👍